src/utilities/theme/utils.ts

```diff
@@ -71,33 +86,49 @@ export function buildColors(theme: ThemeConfig) {
 
   const lightSurface = isLight(hexToRgb(colors.surface));
 
-  const allColors = Object.entries(colorAdjustments).reduce(
-    (accumulator, [colorRole, colorAdjustment]) => {
-      if (colorAdjustment == null) return accumulator;
-
-      const baseColor = hexToHsluvObj(colors[colorAdjustment.baseColor]);
-      const {
-        hue = baseColor.hue,
-        saturation = baseColor.saturation,
-        lightness = baseColor.lightness,
-        alpha = 1,
-      } = colorAdjustment[lightSurface ? 'light' : 'dark'];
-
+  return Object.entries(roleVariants).reduce(
+    (acc1, [role, variants]: [Role, Variant[]]) => {
+      const base = hexToHsluvObj(colors[role]);
       return {
-        ...accumulator,
-        [colorRole]: hslToString({
-          ...colorToHsla(hsluvToHex([hue, saturation, lightness])),
-          alpha,
-        }),
+        ...acc1,
+        ...variants.reduce((acc2, {name, light, dark}) => {
+          const configs: Record<string, HslaSetting> = {
+            default: lightSurface ? light : dark,
+          };
+
+          if (!isEqual(light, dark)) {
+            configs.Inverse = lightSurface ? dark : light;
+            configs.Light = light;
+            configs.Dark = dark;
+          }
+
+          return {
+            ...acc2,
+            ...Object.entries(configs).reduce((acc3, [variant, config]) => {
+              const {
+                hue = base.hue,
+                saturation = base.saturation,
+                lightness = base.lightness,
+                alpha = 1,
+              } = config;
+
+              const displayName =
+                variant === 'default' ? name : `${name}${variant}`;
+
+              return {
+                ...acc3,
+                [displayName]: hslToString({
+                  ...colorToHsla(hsluvToHex([hue, saturation, lightness])),
+                  alpha,
+                }),
+              };
+            }, {}),
+          };
+        }, {}),
```

Comment:
Some of the confusion around these nested reducers could be mitigated by using better names for the accumulators. It would also help understandability while refactoring.

Reply:
Yes, this will mostly go away when the factory is refactored to support child theme providers (only generating variants for a roles provided)

Reply:
Nice 👍